File: Sort/acw787_merge_sort.py
# V1 – 传统归并排序（涉及边界问题）
def merge_sort(arr, l, r):
    if l >= r: return
    mid = l + r >> 1
    merge_sort(arr, l, mid)
    merge_sort(arr, mid + 1, r)
    # merge(arr, start, mid, end)
    i, j, tmp = l, mid + 1, []
    while i <= mid and j <= r:
        if arr[i] <= arr[j]:
            tmp.append(arr[i])
            i += 1
        else:
            tmp.append(arr[j])
            j += 1
    while i <= mid:
        tmp.append(arr[i])
        i += 1
    while j <= r:
        tmp.append(arr[j])
        j += 1
    for k in range(len(tmp)): arr[l + k] = tmp[k]
    return arr


# V2 – 改进版（无需考虑边界问题，但需使用切片）
def merge_sort2(arr):
    if len(arr) > 1:
        mid = 0 + len(arr) >> 1
        L, R = arr[:mid], arr[mid:]
        merge_sort2(L)
        merge_sort2(R)
        # merge
        i, j, k = 0, 0, 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]  # 原地修改（从0开始--∵切片）
                k += 1
                i += 1
            else:
                arr[k] = R[j]
                k += 1
                j += 1
        while i < len(L):
            arr[k] = L[i]  # 原地修改（从0开始--∵切片）
            k += 1
            i += 1
        while j < len(R):
            arr[k] = R[j]
            k += 1
            j += 1
    return arr


if __name__ == "__main__":
    N = int(input())
    arr = list(map(int, input().split()))
    # v1
    sorted_arr = merge_sort(arr, 0, len(arr) - 1)
    for num in sorted_arr:
        print(num, end=' ')
    # v2
    sorted_arr2 = merge_sort2(arr)
    print(' '.join(list(map(str, sorted_arr2))))

File: Sort/acw787_merge_sort.java
package Sort;

import java.util.Scanner;

public class acw787_merge_sort {
    // 稳定算法；非原地排序；好坏均复杂度均为O(nlogn); 自下而上
    public void mergeSort(int[] nums, int start, int end) {
        if (start >= end) return;
        int mid = start + end >> 1;
        mergeSort(nums, start, mid);
        mergeSort(nums, mid+1, end);
        merge(nums, start, mid, end);
    }

    private void merge(int[] nums, int start, int mid, int end) {
        int i = start, j = mid+1, k = 0;
        int n = end - start + 1;
        int[] tmp = new int[n];

        while (i <= mid && j <= end) {
            if (nums[i] < nums[j]) {
                tmp[k++] = nums[i++];
            } else {
                tmp[k++] = nums[j++];
            }
        }
        while (i <= mid) tmp[k++] = nums[i++];
        while (j <= end) tmp[k++] = nums[j++];
        System.arraycopy(tmp, 0, nums, start, n); // ∴?非原地排序，耗内存;
//        for (int p = 0; p < n; p++) {
//            nums[start+p] = tmp[p];
//        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }
        acw787_merge_sort sol = new acw787_merge_sort();
        sol.mergeSort(nums, 0, n-1);
    }
}
